### ISSUE TITLE
Enumerable.Except has a bug

### ISSUE THREAD

**Issue by Iplaylf2** [dotnet/corefx#25541]
```
Int32[] A = new Int32[] { 1, 1 }, B = new Int32[] { 9, 1 };
Boolean bug = A.Except(B).Any() != B.Except(A).Any();//true
```
The question of translation.In Chinese Except was translate to "difference of set"(差集)

**Comment by Clockwork-Muse** [comment]
(for future finders)
Behaving as designed.

Even disregarding the fact that set operations (like `Except`) result in a collection of distinct elements, _all_ matching elements are considered for the operation.  This means that `A.Except(B)` results in an empty set, and `B.Except(A)` results in a set with one element (`9`), with the resulting calls to `Any()` generating opposite boolean results.

**Comment by mladenb** [comment]
TL/DR; Browse the source code of the Intersect/Except and instead of Set.Add() and Set.Remove(), use Set.Contains().

Maybe this was a poor design decision. The arguments of the Intersect/Except methods are IEnumerable, so in general, it's not a collection of unique elements.

If the arguments were of HashSet (or just Set) type, that would guarantee the uniqueness of their elements. In that case it would make sense to return a collection of unique elements (using Set.Add / Set.Remove), and clearly specify it (using the return type of HashSet/Set),

But since the contract of the Intersect/Except didn't specify that their arguments are collections of unique elements, it shouldn't assume that those are. Hence, the implementation for the IEnumerable arguments should have been implemented using Set.Contains, instead of Set.Add / Set.Remove.

Links for the source code: https://github.com/dotnet/corefx/blob/master/src/System.Linq/src/System/Linq/Intersect.cs
https://github.com/dotnet/corefx/blob/master/src/System.Linq/src/System/Linq/Except.cs

**Comment by Clockwork-Muse** [comment]
@mladenb - `Intersect`/`Except` (along with `Union`) are "set operations" [in the mathematical sense](https://en.wikipedia.org/wiki/Set_(mathematics)), where each element will be unique for both input and output.

The reason the parameters aren't `ISet` is to make them more widely usable.  For instance, it makes it much easier to specify a collection literal:
```csharp
var result = someSet.Except(new[] { 1, 4, 10 });
```
...or to take whatever other existing non-`ISet` collection you have laying around to use as input (because the results will be the same whether you have duplicate elements or not).

The reason the operations don't return `ISet` is because LINQ streams results - you don't have all the elements immediately - and the API of `ISet` is completely wrong for that (calling, for example, `ISet.Add(...)` in such a situation would not work at all).

**Comment by mladenb** [comment]
@Clockwork-Muse I partially agree with you. Maybe it would help avoid the confusion (due to different assumptions) if there were additional overloads that don't return just the unique elements, but all matched. Or if there was an Options argument passed to the methods, specifying if the uniqueness is desired or not.
